### PR TITLE
feat: Use IAM database authentication

### PR DESCRIPTION
**Pull request by lamkeewei**
## Problem

Using IAM for database authentication improves security by:
1. Remove the need to regularly rotate our database credentials. Auth tokens are generated dynamically and are only valid for 15 minutes.
2. Better access control to database for different user types can be achieved through the use IAM policies
3. No longer need to pass in database credentials as environment variables

### IAM database authentication
Refer to [document](https://docs.google.com/document/d/18zqublPXLW2R6s2eAjUg92tt2Ycl-TNnEH_B9YYc7gE/edit?usp=sharing) for more implementation details.

### Connection helper script
`connect-rds.sh` helps to simplify/automate the steps necessary to connect to an RDS instance using the developer's AWS credentials. 

#### Overview
Broadly, the script does the following to help connect to an RDS instance from a developer's laptop:
1. Generate temporary AWS credentials using MFA. MFA is required for the actions that will be required later.
2. (Optional) Send local SSH key to the jumphost using EC2 Instance Connect. We will be using the jumphost to establish an SSH tunnel later.
3. Generate an RDS authentication token for use as password when connecting to the RDS instance.
4. Establish an SSH tunnel through the jumphost to our RDS instance. Use the password in step 3 for authentication. 

Usage for the script is as follow:
```
usage: connect-rds.sh [CONFIG_FILE_PATH] [MFA_ARN] [SSH_PRIVATE_KEY] [SSH_PUBLIC_KEY]

Arguments:
  CONFIG_FILE_PATH  - path to config file with connection variables
  MFA_ARN           - ARN for MFA device. Can be retrieved from AWS Console > My Security Credentials.
  SSH_PRIVATE_KEY   - SSH private key to use for connecting to the jumphost. Can be generated with ssh-keygen.
  SSH_PUBLIC_KEY    - Optional SSH public key to use for connecting to the jumphost. Can be generated with ssh-keygen.
```

#### Try it out
To give this a test run, make sure the following is completed first: 
1. Download the config file in the team's 1Password (Staging Connect RDS Config). This file contains key value pairs (RDS host, port, EC2 instance id, etc) required to connect to the RDS instance. You can also use the EC2 instance connect version to test out.
2. Make sure your AWS CLI is installed and configured with your credentials. Run `aws configure`
3. Ensure that you have an SSH key pair ready (else user `ssh-keygen`)

When launched, you will be prompted for your MFA OTP. Upon successfully creating the SSH tunnel, connection details will be provided. After which, you can use your preferred client to connect to the RDS instance (using localhost).
```
Enter your MFA token: 111111
Loaded config file located at ~/rds-staging.conf
Generated AWS session credentials
Sent SSH key (/path/to/your/public/key) to jumphost using EC2 instance connect

Created SSH tunnel to RDS instance. Connect using the following credentials:
  * host: localhost
  * port: 15432
  * username: [REDACTED]
  * password: [REDACTED]

Press (Ctrl-C) to close the tunnel
```

## Tests
_For the purpose of testing, a `db.t2.micro` RDS instance is created in our staging environment. Slower performance is expected when testing._

**Connection script**
- Download `rds-staging.conf` from 1Password.
- Follow the instructions above
- Verify that you are able to connect to the database with the generated credentials
- Try the same steps again without using EC2 instance connect

**IAM database connection**
- Set `STAGING_DB_USE_IAM` to `true` on Travis
- Update the following database connection URIs in Travis. For easy copy-and-pasting, look for "RDS IAM Test DB URIs" in our 1Password. Use the `iam` field value.
   - `STAGING_EMAIL_DB_URI `
   - `STAGING_TELEGRAM_HANDLER_DB_URI `
   - `STAGING_TWILIO_DB_URI `
   - `STAGING_UNSUBSCRIBE_DIGEST_DB_READ_REPLICA_URI `
   - `STAGING_UNSUBSCRIBE_DIGEST_DB_URI `
- Update `DB_URI` and `DB_READ_REPLICA_URI` for Elastic Beanstalk
- Update `DB_URI` in `staging-sending` and `staging-logging` task definitions
- Trigger a deploy
- Test out all channels to verify that all functionality work as expected

## Deploy Notes
Detailed checklist and contingencies for deployment will be in postman internal docs repo.

**New environment variables**:
- `DB_USE_IAM` : Whether to use IAM for authentication to database

**New scripts**:
- `connect-rds.sh` : Automate SSH tunnel with EC2 instance connect to RDS instance

**New dependencies**:
- `aws-sdk` : Added to serverless code

## TODOs
- [x] Investigate downtime incurred for enabling IAM authentication
   - On test instance, this takes < 5 minutes
- [x] Create and attach all staging IAM policies
- [x] Test on staging without IAM authentication
- [x] Test on staging using RDS test instance with IAM authentication
- [x] Investigate if we can disable password auth for RDS master user
   - Once `rds_iam` role has been granted to a previously password authenticated user, IAM authentication will be required. The record for IAM authentication (`hostssl all +rds_iam 0.0.0.0/0 pam`) has a higher precedence in `pg_hba.conf`. 
- [x] Create checklist and rehearse setup on staging database before merge into `develop`

### CODE REVIEW

**Comment by lshaowei18** [comment]
Once we setup our Secrets Manager properly, do you think we should store the `rds-staging-instance-connect.conf ` in Secrets Manager and use AWS CLI to fetch it?

**Comment by lamkeewei** [comment]
> Do you think we should store the `rds-staging-instance-connect.conf ` in Secrets Manager and use AWS CLI to fetch it?

Hmm...tbh there are no real secrets (mostly just hostnames, usernames, config flags) in the conf file. It's mainly for convenience and avoiding the need to pass in so many environment variables. I think whichever is the easiest mode of distribution of the file to our team should suffice.

File: backend/src/core/utils/rds-iam.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import AWS from 'aws-sdk'
+import { Config } from 'sequelize'
+
+import config from '@core/config'
+import { configureEndpoint } from '@core/utils/aws-endpoint'
+
+const rdsSigner = new AWS.RDS.Signer(configureEndpoint(config))
+
+export type MutableConfig = {
+  -readonly [K in keyof Config]: Config[K]
```
Comment (lshaowei18) [inline backend/src/core/utils/rds-iam.ts:10]: Hmmm whats the goal for using -readonly?

**Comment by lshaowei18** [comment]
`connect-rds.sh`
- [x] EC2 Instance Connect
- [x] SSH key

File: backend/src/core/utils/rds-iam.ts (hunk 1 above)
Comment (lamkeewei) [reply backend/src/core/utils/rds-iam.ts:10]: The `password` field in the exported `Config` type from `sequelize` is [by default readonly](https://github.com/sequelize/sequelize/blob/81c33d8c2fb719409bf5692eb6a50a5883c2b6f3/types/lib/sequelize.d.ts#L144). However, we need to set it to the generated auth token in the `beforeConnect` hook. Hence, I decided to recast `Config` to a custom `MutableConfig` type where all fields are mutable. Alternatively, we could cast to `any` but decided to do this to preserve the type information.

**Comment by lshaowei18** [comment]
lgmt, let me know if you need any help with testing since there is so many different lambdas and functionality that might be affected.

**Review by lshaowei18 — APPROVED** [review]

**Pull request merged**